backend/SyncReverseProxy/ProxyKernel.cs

```diff
@@ -60,27 +63,74 @@ public static class ProxyKernel
+        app.Map("/api/v03/{**catch-all}",
+            async (IOptions<HgConfig> hgConfig,
+                HttpContext context,
+                [FromQuery(Name = "repoId")] string projectCode) =>
             {
-                var projectCode = context.Request.GetProjectCode();
-                if (projectCode is not null)
+                await Forward(context, httpClient, hgConfig.Value.HgResumableUrl, projectCode);
+            }).RequireAuthorization(authorizeAttribute).WithMetadata(HgType.resumable);
+
+        //hgweb
+        app.Map($"/{{{ProxyConstants.HgProjectCodeRouteKey}}}/{{**catch-all}}",
+            async (IOptions<HgConfig> hgConfig,
+                HttpContext context,
+                [FromRoute(Name = "project-code")] string projectCode) =>
+            {
+                await Forward(context, httpClient, hgConfig.Value.HgWebUrl, projectCode);
+            }).RequireAuthorization(authorizeAttribute).WithMetadata(HgType.hgWeb);
+        app.Map($"/hg/{{{ProxyConstants.HgProjectCodeRouteKey}}}/{{**catch-all}}",
+            async (IOptions<HgConfig> hgConfig,
+                HttpContext context,
+                [FromRoute(Name = ProxyConstants.HgProjectCodeRouteKey)] string projectCode) =>
+            {
+                var hgWebUrl = hgConfig.Value.HgWebUrl;
+                if (hgWebUrl.EndsWith("hg/"))
                 {
-                    Activity.Current?.AddTag("app.project_code", projectCode);
+                    hgWebUrl = hgWebUrl[..^"hg/".Length];
                 }
 
-                var eventsService = context.RequestServices.GetRequiredService<ProxyEventsService>();
-                var proxyFeature = context.Features.Get<IReverseProxyFeature>();
-                await next(context);
-                ArgumentNullException.ThrowIfNull(proxyFeature);
-                await eventsService.AfterRequest(context, proxyFeature);
-            }))
-            .RequireAuthorization(new AuthorizeAttribute
-            {
-                AuthenticationSchemes = string.Join(',', BasicAuthHandler.AuthScheme, extraAuthScheme ?? "")
-            });
+                await Forward(context, httpClient, hgWebUrl, projectCode);
+            }).RequireAuthorization(authorizeAttribute).WithMetadata(HgType.hgWeb);
+    }
+
+    private enum HgType
+    {
+        hgWeb,
+        resumable
+    }
+
+    private static async Task Forward(HttpContext context,
+        HttpMessageInvoker httpClient,
+        string destinationPrefix,
+        string projectCode)
+    {
+        var forwarder = context.RequestServices.GetRequiredService<IHttpForwarder>();
+        var eventsService = context.RequestServices.GetRequiredService<ProxyEventsService>();
+        Activity.Current?.AddTag("app.project_code", projectCode);
+        await forwarder.SendAsync(context, destinationPrefix, httpClient);
+        var hgType = context.GetEndpoint()?.Metadata.OfType<HgType>().FirstOrDefault();
+        if (hgType == HgType.hgWeb)
```

Comment:
Just out of curiosity, why do we not update the project-last-change here for resumable? Is it because we can't tell if the sync wrote to the repo? Or...

Reply:
I didn't actually determine what api calls precipitate resumable changing project data so I didn't implement it. I'll make an issue for that.